### PR TITLE
WIP: Simpler SubArrays

### PR DESCRIPTION
**Pull request by mbauman**
This is a simplification I've long wanted to perform.  The change in field name from `indexes` -> `indices` is a breaking point… and as long as we're breaking the layout of SubArrays, let's make them right.

In short, `SubArray`s now only have two fields: the parent array and the tuple of indices.  Linearly-compatible SubArrays are supported by collapsing multiple indices down into a single `AbstractRange` or a `ReshapedRange`.  Amazingly, my examinations of the performance here look to be on-par or better, with only a handful of tests showing slight (10-20%) regressions.  I'll need to go back and make sure those are noise or see if I can fix their performance.  Better performance comes from allowing fewer indirections — we don't really need to reshape the parent arrays in many cases — and enabling more SubArrays to be considered IndexLinear.

Even better performance could be had by allowing `reshape` of these linear-ly compatible SubArrays to _remain_ as subarrays.  This breaks a number of tests that are specifically expecting `reshape(::SubArray)` to always return a `ReshapedArray` — @timholy would you be okay breaking that assumption?

### CODE REVIEW

File: base/subarray.jl [hunk 1]
```diff
@@ -82,25 +95,29 @@ julia> parent(s_a)
 ```
 """
 parent(V::SubArray) = V.parent
-parentindexes(V::SubArray) = V.indexes
+parentindexes(V::SubArray) = V.indices
```
Comment (timholy) [inline base/subarray.jl]: This seems increasingly weird as an exported function. Should we just get rid of it?

File: base/subarray.jl [hunk 2]
```diff
@@ -1,60 +1,73 @@
...
-struct SubArray{T,N,P,I,L} <: AbstractArray{T,N}
+# The view itself supports fast linear indexing if length(indices) == 1 and IndexStyle(indices[1]) is LinearFast()
+"""
+    SubArray{T,N,P,I} <: AbstractArray{T,N}
+
+A specialized array type for that represents a view into an indexed subsection of a "parent" array.
+
+Typically constructed with [`view`](@ref) or indexing inside a [`@view`](@ref) or [`@views`](@ref) macro,
+SubArrays simply hold onto the parent array and the indices. The additional type parameters concretely
+identify the type of the parent array (`P`) and the type of the tuple of indices (`I`).
```
Comment (timholy) [inline base/subarray.jl]: is it always a "tuple of indices" now?

File: base/subarray.jl [hunk 3]
```diff
@@ -1,60 +1,73 @@
...
 end
 
-check_parent_index_match(parent, indexes) = check_parent_index_match(parent, index_ndims(indexes...))
-check_parent_index_match(parent::AbstractArray{T,N}, ::NTuple{N, Bool}) where {T,N} = nothing
-check_parent_index_match(parent, ::NTuple{N, Bool}) where {N} =
-    throw(ArgumentError("number of indices ($N) must match the parent dimensionality ($(ndims(parent)))"))
+# StridedSubArrays depend upon Ranges wrapped in ReshapedArrays, so more functionality
+# is defined within reshapedarray.jl
+
+# Collapse a set of linearly compatible indices into a single index
```
Comment (timholy) [inline base/subarray.jl]: Some brief docs/comments might be helpful: extracting what's going on by looking at the code is less straightforward than the idea itself.

Presumably the input here must be de-colonified? I've lost track of which function is in charge of that.

File: base/subarray.jl [hunk 4]
```diff
@@ -273,101 +272,26 @@ compute_stride1(s, inds, I::Tuple{Any, Vararg{Any}}) = throw(ArgumentError("inva
...
-            dimindex[j] = i
-            j += 1
-        end
-    end
-    dimindex
-end
+# Most of the time, reshape needs to add a full layer of indirection. But when
+# there's only one index then we just need to reshape that index
+# reshape(V::OneIndexSubArray, dims::Dims) =
+#     SubArray(V.parent, (reshape(V.indices[1], dims),))
```
Comment (timholy) [inline base/subarray.jl]: delete?

**Review by timholy — COMMENTED** [review]
I really like this: I think it's a much more straightforward representation. Interesting how our oldest "view" has now benefited from one of the newest.

Performance would have been my main worry, so it will be great to see the benchmarks (including any you think need to be added). But it sounds like those are in good shape, too. :100:

**Comment by timholy** [comment]
> @timholy would you be okay breaking that assumption?

Yes. Perhaps I added those tests myself, but in general I think that kind of test emphasizes implementation details rather than some kind of behavior.

File: base/subarray.jl (hunk 2 above)
Comment (mbauman) [reply base/subarray.jl]: Yup!

File: base/subarray.jl (hunk 3 above)
Comment (mbauman) [reply base/subarray.jl]: Yes, I'll definitely need to revise the docs/devdocs.

Colon conversion is now done within `to_indices`.

File: base/subarray.jl (hunk 1 above)
Comment (mbauman) [reply base/subarray.jl]: I definitely agree! I had it on the chopping block but a handful of packages use it a surprising amount of times. I'll want to see how they use it first.

**Comment by mbauman** [comment]
Thanks Tim! I'll charge ahead here as I have time. Nanosoldier has a bunch of SubArray specific additions, so once it's feeling better we can get some numbers from it.

**Comment by mbauman** [comment]
Well, shoot.  I'm afraid this has hit a dead-end.

It's very common to create a 2-d view like `V = view(rand(5,5), :, 2:4)`, for example.  Those indices are linearly compatible, so this PR collapses them down to a single `reshape(6:20, 5, 3)`.  That works great and is very efficient.

The problem is that it's common to then subsequently look at views of `V` — like `view(V, 2:4, 1:2)`.  Those indices aren't linearly compatible, but they are strided (and therefore BLAS-compatible).  `SubArray` currently simply re-indexes the slice by `2:4`, and `(2:4)[1:2]` to get its new set of indices — `(2:4, 2:3)`.

But with this PR, we no longer know that `V` was originally constructed with the indices `(:, 2:4)`.  All we know is that there's one `reshape(::UnitRange, 5, 3)`.  When SubArray goes to re-index that one reshaped range, it's not able to back out the fact that `reshape(6:20, 5, 3)[2:4, 1:2]` could be represented _in this case_ as the pair `(2:4, 2:3)`.  It's not trivial to compute, nor could it be type-stable — a user could just as easily use `reshape(7:21, 5, 3)` as an index, but that does not align with the strides of its parent and therefore no such pair of indices exist.

Now, we could add that information back into a more specialized reshaped range type, say, by also holding onto the original indices in addition to the reshaped range… but at that point we've effectively worked our way back to the status quo.

This PR does much better at reshaping linear SubArrays… but in practice it's much more common to iteratively subset an array than it is to reshape it.  Thanks, comprehensive LinAlg test suite!

A number of these optimizations could still apply, and we still need to rename that field, but I'll tackle all that separately.

**Pull request closed without merging**